File: leetcode/11-container-with-most-water.py
# https://leetcode.com/problems/container-with-most-water/ (top-interview-questions)

from typing import List

class Solution:
	def maxArea(self, height: List[int]) -> int: # Time : O(n)
		area, left, right = 0, 0, len(height) - 1

		while left < right:
			distance = right - left
			currArea = 0
			if height[left] > height[right]:
				currArea = distance * height[right]
				right -= 1
			else:
				currArea = distance * height[left]
				left += 1
			area = max(currArea, area)
		return area

File: leetcode/11-container-with-most-water.js
// https://leetcode.com/problems/container-with-most-water/ (top-interview-questions)

/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
  let maxWater = 0;
  let l = 0;
  let r = height.length - 1;

  while (l < r) {
    // shortest line * distance between the two lines
    const waterAmount = Math.min(height[l], height[r]) * (r - l);
    // update the max water amount
    maxWater = Math.max(maxWater, waterAmount);
    // pointer update (move the shortest line)
    if (height[l] < height[r]) {
      l += 1;
    } else {
      r -= 1;
    }
  }

  return maxWater;
};